### PR TITLE
Change protocol select to flatbuffer

### PR DESCRIPTION
**Pull request by driftluo**
Change protocol select to flatbuffer

### CODE REVIEW

File: src/protocol_select/mod.rs [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+            let mut versions: Vec<String> = Vec::new();
+            let fbs_versions = fbs_protocol_message.support_versions().unwrap();
+            for i in 0..fbs_versions.len() {
+                versions.push(fbs_versions.get(i).to_owned());
+            }
+            Ok(ProtocolMessage {
+                name: fbs_protocol_message.name().unwrap().to_owned(),
+                support_versions: versions,
+            })
+        }
```
Comment (TheWaWaR) [inline src/protocol_select/mod.rs]: Use `match` can avoid `unwrap()` calls

File: src/protocol_select/mod.rs [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+                .send(Bytes::from(message.encode()))
+                .from_err()
+                .map(|socket| socket)
+        })
+        .and_then(|socket| {
+            socket
+                .into_future()
+                .map_err(|(e, socket)| {
+                    let _ = socket.into_inner().shutdown();
+                    e
```
Comment (TheWaWaR) [inline src/protocol_select/mod.rs]: Why `shutdown()` here?

File: src/session.rs [hunk 3]
```diff
@@ -220,25 +240,42 @@ where
...
+            .map(|proto_meta| {
+                let name = proto_meta.name();
+                let proto_msg = ProtocolMessage::new(&name, proto_meta.support_versions());
+                (name, proto_msg)
+            })
+            .collect::<HashMap<String, ProtocolMessage>>();
+
+        let task = server_select(sub_stream, proto_metas)
+            .and_then(|(mut handle, name, version)| {
+                if version.is_some() {
```
Comment (TheWaWaR) [inline src/session.rs]: Better `match version`

File: src/protocol_select/mod.rs (hunk 2 above)
Comment (driftluo) [reply src/protocol_select/mod.rs]: Any error occurs during the handshake, must close the sub stream

**Pull request merged**